source/common/event/scaled_timer.cc

```diff
@@ -0,0 +1,207 @@
+  const TimerCb callback_;
+  absl::optional<ActiveTimer> active_;
+};
+
+ScaledTimerManager::ScaledTimerImpl::ScaledTimerImpl(ScaledTimerManager& manager, TimerCb callback)
+    : manager_(manager), callback_(std::move(callback)) {}
+
+ScaledTimerManager::ScaledTimerImpl::~ScaledTimerImpl() {
+  if (active_.has_value()) {
+    manager_.descheduleTimer(this);
+  }
+}
+
+void ScaledTimerManager::ScaledTimerImpl::enableTimer(const std::chrono::milliseconds& delay,
+                                                      const ScopeTrackedObject* object) {
+  if (active_.has_value()) {
+    disableTimer();
+  }
+  active_.emplace(ActiveTimer{
+      .start = manager_.dispatcher_.timeSource().monotonicTime(),
+      .delay = delay,
+      .object = object,
+  });
+  manager_.scheduleTimer(this);
+}
+
+void ScaledTimerManager::ScaledTimerImpl::enableHRTimer(const std::chrono::microseconds& delay,
+                                                        const ScopeTrackedObject* object) {
+  if (active_.has_value()) {
+    disableTimer();
+  }
+  active_.emplace(ActiveTimer{
+      .start = manager_.dispatcher_.timeSource().monotonicTime(),
+      .delay = delay,
+      .object = object,
+  });
+  manager_.scheduleTimer(this);
+}
+
+void ScaledTimerManager::ScaledTimerImpl::disableTimer() {
+  manager_.descheduleTimer(this);
+  active_.reset();
+}
+
+void ScaledTimerManager::ScaledTimerImpl::dispatch(Dispatcher& dispatcher) {
+  ASSERT(active_.has_value());
+  if (active_->object == nullptr) {
+    callback_();
+  } else {
+    ScopeTrackerScopeState scope(active_->object, dispatcher);
+    callback_();
+  }
+  active_.reset();
+}
+
+const absl::optional<ScaledTimerManager::ScaledTimerImpl::ActiveTimer>&
+ScaledTimerManager::ScaledTimerImpl::active() const {
+  return active_;
+}
+
+ScaledTimerManager::ScaledTimerManager(Dispatcher& dispatcher)
+    : dispatcher_(dispatcher),
+      dispatch_timer_(dispatcher.createTimer([this] { dispatchCallback(); })), timers_(1.0) {}
+
+Event::TimerPtr ScaledTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<ScaledTimerImpl>(*this, std::move(callback));
+}
+
+void ScaledTimerManager::setDurationScaleFactor(double scale_factor) {
+  timers_.setScaleFactor(scale_factor);
+  internalScheduleTimer();
+}
+
+ScaledTimerManager::ScaledTimerList::ScaledTimerList(double scale_factor)
+    : timers_(Compare{.scale_factor = scale_factor}) {}
+
+void ScaledTimerManager::ScaledTimerList::addActive(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.insert(timer);
+}
+
+void ScaledTimerManager::ScaledTimerList::removeActive(ScaledTimerImpl* timer) {
+  ASSERT(timer->active().has_value());
+  timers_.erase(timer);
+}
+
+void ScaledTimerManager::ScaledTimerList::setScaleFactor(double scale_factor) {
+  std::set<ScaledTimerImpl*, Compare> reordered(timers_.begin(), timers_.end(),
+                                                Compare{.scale_factor = scale_factor});
```

Comment:
Doing a linear scan and O(nlgn) set insertion computation to reorder entries when the scale factor changes seems very expensive.  I wonder what alternate approaches we could consider to avoid this expense.